File: scala/queen-attack/src/main/scala/Queens.scala
import Queens._
import Chessboard._

object Queens {

  // needed for the tests to compile and run
  def apply() = this

  type Queen = Option[Position]

  def boardString(whiteQueen: Queen, blackQueen: Queen): Chessboard = {
    def getSquare(position: Position): Square = {
      def queenSquare(queen: Queen, square: Square): Option[Square] =
        queen filter (_ == position) map (_ => square)

      queenSquare(whiteQueen, WhiteQueenSquare) orElse
      queenSquare(blackQueen, BlackQueenSquare) getOrElse EmptySquare
    }

    val squares: Squares =
      for {
        row <- Rows
        col <- Columns
        square = getSquare(Position(row, col))
      } yield square

    val rows: Seq[Squares] = squares.grouped(BoardSize).toSeq
    val rowsAsStrings: Seq[String] = rows map (_.mkString("", " ", "\n"))
    rowsAsStrings mkString
  }

  def canAttack(position1: Position, position2: Position): Boolean = {
    val deltaRow = math.abs(position1.row - position2.row)
    val deltaCol = math.abs(position1.column - position2.column)

    def onSameRow = deltaRow == 0
    def onSameCol = deltaCol == 0
    def onSameDiagonal = deltaRow == deltaCol

    onSameRow || onSameCol|| onSameDiagonal
  }
}

case class Position(row: Row, column: Column)

object Chessboard {
  type Chessboard = String
  type Square = Char
  type Squares = Seq[Square]

  type Row = Int
  type Column = Int

  val BoardSize = 8

  val Rows = (0 until BoardSize)
  val Columns = (0 until BoardSize)

  val EmptySquare: Square = '_'
  val WhiteQueenSquare: Square = 'W'
  val BlackQueenSquare: Square = 'B'
}

File: haskell/queen-attack/Queens.hs
module Queens (boardString, canAttack) where

import Data.Maybe (catMaybes)
import Control.Monad (mfilter)
import Data.List (intersperse)
import Data.List.Split (chunksOf)

type Row = Int
type Col = Int
type Position = (Row, Col)
type Queen = Maybe Position
type Chessboard = String
type Square = Char

boardString :: Queen -> Queen -> Chessboard
boardString whiteQueen blackQueen =
  concat rows
  where
    rows = map addSeparators $ chunksOf 8 boardSquares
    boardSquares = fmap getSquare boardPositions
    boardPositions = [ (row,col) | row <- [0..7], col <- [0..7] ]
    addSeparators = flip (++) "\n" . intersperse ' '

    getSquare :: Position -> Square
    getSquare position =
      head $ catMaybes [queenSquare 'W' whiteQueen, queenSquare 'B' blackQueen, Just '_']
      where
        queenSquare :: Square -> Queen -> Maybe Square
        queenSquare square = fmap (const square) . mfilter (== position)

canAttack :: Position -> Position -> Bool
canAttack position1 position2 =
  onSameRow || onSameCol|| onSameDiagonal
  where
    onSameRow = deltaRow == 0
    onSameCol = deltaCol == 0
    onSameDiagonal = deltaRow == deltaCol
    deltaRow = delta fst
    deltaCol = delta snd
    delta f = abs (f position1 - f position2)
